java/src/main/java/org/eclipse/ditto/client/messaging/internal/Retry.java

```diff
@@ -0,0 +1,202 @@
+ * In this case this action must return a result which also can be null.
+ * As soon as a result is returned is by the supplier, the action is considered to be finished successfully.
+ *
+ * @param <T> Type of the result
+ */
+final class Retry<T> implements Supplier<CompletionStage<T>> {
+
+    private static final int[] TIME_TO_WAIT_BETWEEN_RETRIES_IN_SECONDS = new int[]{1, 1, 2, 3, 5, 8, 13};
+    private static final Logger LOGGER = LoggerFactory.getLogger(Retry.class.getName());
+
+    private final String sessionId;
+    private final String nameOfAction;
+    private final Supplier<T> retriedSupplier;
+    private final ExecutorService executorService;
+
+
+    private Retry(final String nameOfAction,
+            final String sessionId,
+            final Supplier<T> retriedSupplier,
+            final ExecutorService executorService) {
+
+        this.sessionId = sessionId;
+        this.nameOfAction = nameOfAction;
+        this.retriedSupplier = retriedSupplier;
+        this.executorService = executorService;
+    }
+
+    private static int ensureIndexIntoTimeToWaitBounds(final int index) {
+        if (index < 0) {
+            return 0;
+        } else if (index >= TIME_TO_WAIT_BETWEEN_RETRIES_IN_SECONDS.length) {
+            return TIME_TO_WAIT_BETWEEN_RETRIES_IN_SECONDS.length - 1;
+        } else {
+            return index;
+        }
+    }
+
+    /**
+     * Executes the provided supplier until a result is returned.
+     *
+     * @return A completion stage which finally completes with the result of the supplier. Result can be null.
+     */
+    @Override
+    public CompletionStage<T> get() {
+        return CompletableFuture.supplyAsync(() -> this.doGet(1), executorService);
+    }
+
+    private T doGet(int attempt) {
+        try {
+            return retriedSupplier.get();
+        } catch (final RuntimeException e) {
+            // log error, but try again (don't end loop)
+            LOGGER.error("Client <{}>: Failed to <{}>: {}.", sessionId, nameOfAction, e.getMessage());
+            waitFor(getTimeToWaitForAttempt(attempt));
+            return doGet(attempt + 1);
+        }
+    }
+
+    private Duration getTimeToWaitForAttempt(int attempt) {
+        final int attemptIndex = ensureIndexIntoTimeToWaitBounds(attempt - 1);
+        return Duration.ofSeconds(TIME_TO_WAIT_BETWEEN_RETRIES_IN_SECONDS[attemptIndex]);
+    }
+
+    private void waitFor(final Duration timeToWait) {
+        try {
+            LOGGER.info("Client <{}>: Waiting for <{}> second(s) before retrying to <{}>.",
+                    sessionId, timeToWait.getSeconds(), nameOfAction);
+            TimeUnit.SECONDS.sleep(timeToWait.getSeconds());
+        } catch (final InterruptedException ie) {
+            LOGGER.error("Client <{}>: Interrupted while waiting for the next try to <{}>.",
+                    sessionId, nameOfAction, ie);
+            Thread.currentThread().interrupt();
```

Comment:
This will become dangerous later, because we will eventually let users bring their own thread pool for the providers so that they can start thousands of Ditto clients in 1 JVM. How about changing the type of `this.callbackExecutor` to `ScheduledExecutorService` and use its scheduling function?

Reply:
Provided updated way of handling the loop which should solve both of your change requests.